Updated crash site /restart command so that guests and regulars can call it when map is cleared

Jayefuu opened the pull request:
Updated crash site /restart command so that guests and regulars can call it when map is cleared
- Only admin can abort, once the map is cleared.
- Guests and Regular can both call /restart when all enemy units and structures are dead
- Tested on S10 by running this command:
/c local surface=game.player.surface
for key, entity in pairs(surface.find_entities_filtered({force="enemy"})) do
	entity.destroy()
end
- Need to merge into develop to test on live server in case conditions at the end of the game are different
- Thought about adding rocket launched condition but decided against it.

Jayefuu committed:
Updated crash site /restart command so that guests and regulars can call it when map is cleared

Updated crash site /restart command so that guests and regulars can call it when map is cleared
- Only admin can abort, once the map is cleared.
- Guests and Regular can both call /restart when all enemy units and structures are dead
- Tested on S10 by running this command:
/c local surface=game.player.surface
for key, entity in pairs(surface.find_entities_filtered({force="enemy"})) do
	entity.destroy()
end
- Need to merge into develop to test on live server in case conditions at the end of the game are different
- Thought about adding rocket launched condition but decided against it.

grilledham commented on map_gen/maps/crash_site/commands.lua:
```diff
@@ -55,6 +57,16 @@ local function restart(args, player)
+        local enemy_count = game.player.surface.count_entities_filtered{force= "enemy"}
+        
+        if (enemy_count ~= 0) then
+            game.player.print('All enemy spawners, worms, buildings, biters and spitters must be cleared for non-admin restart. Current enemy count: '.. enemy_count)
```
I wonder if we should drop the `enemy_count` part. People might use this as a way to count remainging biters, which I'm not sure we want. Also if we drop the count we can put `limit = 1` in the `count_entities_filtered` which makes it quicker.

Jayefuu replied:
Ok. I had it there for debug, I thought it would be useful. But if there's a problem I need to debug I can always print the count with a console command. Will remove the count print.

Jayefuu replied:
I intentionally left it so that regulars/guests can't abort so that I didn't have to handle it. I can't imagine it will matter if the biters are cleared that someone calls restart. If others have the ability to abort a restart then they might fight over it.

grilledham replied:
I think that is fine then.

grilledham commented on map_gen/maps/crash_site/commands.lua:
```diff
@@ -55,6 +57,16 @@ local function restart(args, player)
+        local enemy_count = game.player.surface.count_entities_filtered{force= "enemy"}
+        
+        if (enemy_count ~= 0) then
+            game.player.print('All enemy spawners, worms, buildings, biters and spitters must be cleared for non-admin restart. Current enemy count: '.. enemy_count)
```
Is this meant to be `player.print` instead of `game.player.print`?

Jayefuu replied:
Isn't it the same thing?

grilledham replied:
`game.player` is the player that ran the command, this is `nil` when run from the server console. Which why we have the `player = player or server_player` above.

Now that I think about this some more, `game.player.surface` wont work from the server console. You will need to do something like:
```lua
local RS = require 'map_gen.shared.redmew_surface'
...
RS.get_surface()
```

grilledham commented on map_gen/maps/crash_site/commands.lua:
```diff
@@ -55,6 +57,16 @@ local function restart(args, player)
 
+    if Rank.less_than(player.name, Ranks.admin) then
+        -- Check enemy count
+        local enemy_count = game.player.surface.count_entities_filtered{force= "enemy"}
```
This will include walls, which might be undesirable, not all walls are capturable. But maybe this isn't a problem?

Jayefuu replied:
I'm not sure. I did some play testing and used the destroy all command to allow guests and regulars to do the restart. This worked. But we didn't play through to a natural completion, just an artificial one. We should merge it with develop and if needs be debug the enemy_count limit later. I'm sure there are some things that might change. If the /restart command doesn't work for them one of us can hop on and see what entity remains that's stopping the restart.

grilledham replied:
I guess we can wait to see if this turns into an actual issue.

grilledham commented on map_gen/maps/crash_site/commands.lua:
```diff
@@ -67,7 +79,7 @@ local function restart(args, player)
     print('Abort restart with /abort')
-
+    
     Task.set_timeout_in_ticks(60, callback, {name = player.name, scenario_name = args.scenario_name, state = 10})
```
`scenario_name` could be any scenario name not just crashsite maps. This wasn't a problem when only admins were running the command, but could be now. We don't want troll players starting a diggy map instead.

grilledham replied:
Maybe for non admins we set `args.scenario_name` to `config.scenario_name` below. Would mean moving this function into the Public.control function.

grilledham commented on map_gen/maps/crash_site/commands.lua:
```diff
@@ -124,7 +136,7 @@ function Public.control(config)
             arguments = {'scenario_name'},
             default_values = {scenario_name = default_name},
-            required_rank = Ranks.admin,
+            required_rank = Ranks.guest,
```
This should probably be `auto_trusted`. Easily allows us to remove the ability from players who might be miss using it.

grilledham reviewed (changes requested):
Non admin restarts cannot be aborted by non admins. Is this intentional? If we do allow non admins to abort, then we have to think about non admins aborting when an admin orders the restart.

Have you done a test run on the server, just as a sanity check?

This is a good start, thanks for taking the time to do this. However the nature of this command where it allows any player to perform an action on the server means we have to be careful with it. So there are still some issues we might need to think about.

grilledham commented:
I wonder if it's worth putting a map time check (>5 minutes?) in for non admin users, probably not needed?

I bet once we add this command we will get requests to allow restarts on lose conditions, or some kind of vote restart system.

Jayefuu committed:
Update commands.lua

Jayefuu commented:
@grilledham I have just managed to catch the end of a crash site game. There are 544 enemy entities left, mainly walls and power poles. 

I moved it here in case we want to put it on S10 for debugging:
https://redmew.com/saves?directory=public/final&file=2020-09-28_crash_site_finished.zip

In your opinion should we:
a) Make the players clear all this crap, which could be laborious, to get an enemy count of 0 before they're allowed to reset
b) Filter the count_entities_filtered() to only be units, spawners and turrets, which might have performance impacts?
c) Set a limit so they can restart when less than 500 enemy entities on the map. I checked and immediately upon restarting there are already 1000 entities, so this won't allow them to restart straight away if they don't like the mapgen. And that will grow very quickly. So this could be a way to allow them to restart without clearing ALL walls and power poles without the performance hit of b?

Which option do you prefer?

grilledham commented:
@Jayefuu That is good information to have.

I would do a variation of b).

```lua
local entities_to_check = {
    'biter-spawner',
    ...
    --spawners, biters, spitters, worms, turrets, drones
}

local function map_cleared(force)
    local get_entity_count = force.get_entity_count
    for i = 1, #entities_to_check do
        local name = entities_to_check[i]
        if get_entity_count(name) > 0 then
            return false
        end
    end

    return true
end
```

Jayefuu committed:
Update commands.lua

- Changed behaviour for non-admins. Now checks for spawners, biters and turrets but doesn't require walls and power structures to be destroyed for restart
- Added checks to limit non-admins from changing the scenario by using an argument after /restart. Now defaults to current scenario for non-admins.

grilledham commented on map_gen/maps/crash_site/commands.lua:
```diff
@@ -47,14 +53,46 @@ callback =
+    'spitter-spawner','biter-spawner',
+    'small-worm-turret', 'medium-worm-turret','big-worm-turret', 'behemoth-worm-turret', 
+    'small-spitter', 'medium-spitter', 'big-spitter', 'behemoth-spitter', 
+    'small-biter', 'medium-biter', 'big-biter', 'behemoth-biter'
```
Missing outpost turrets: 'gun-turret', 'laser-turret', 'artillery-turret', 'flamethrower-turret'.
Also maybe combat-robot: 'defender', 'destroyer', 'distractor'

grilledham commented on map_gen/maps/crash_site/commands.lua:
```diff
@@ -47,14 +53,46 @@ callback =
         return
     end
 
+    if Rank.less_than(player.name, Ranks.admin) then
```
I think this should be something like.
```suggestion
    if player ~= server_player and Rank.less_than(player.name, Ranks.admin) then
```
For the case of running this from the server console.

grilledham commented on map_gen/maps/crash_site/commands.lua:
```diff
@@ -47,14 +53,46 @@ callback =
+    if Rank.less_than(player.name, Ranks.admin) then
+        -- Check enemy count      
+        if not map_cleared() then
+            game.player.print('All enemy spawners, worms, buildings, biters and spitters must be cleared for non-admin restart.')
```
Could be `player` instead of `game.player` for consistency with above.

grilledham reviewed (changes requested):
A few minor changes and I think this is ready. 😃

grilledham commented:
You will also have to fix the luacheck warnings.

Jayefuu committed:
Update commands.lua

- Fixed LUA check failures
- Added suggestions by grilledham
- Tested /restart still works from server console

grilledham committed:
Fix non-admin restart

- Stop the entities count from counting entities on the 'nauvis' and 'cutscene' surfaces.
- Only print abort message to admins.
